Add --logfile option and trace all output to this file

Sometimes `paket restore --references-files` fails with code 1 and it's impossible to understand what's went wrong. Build server log contains the following:

```
"...aproject.csproj" (Rebuild target) (46:2) ->
[12:43:07][Step 3/3]        (RestorePackages target) -> 
[12:43:07][Step 3/3]          ...\.paket\paket.targets(36,5): error MSB3073: The command ""...\.paket\paket.exe" restore --references-files "...aproject\paket.references"" exited with code 1. 
```

It would be great if Paket writes error message / full stack trace into stdout (or whatever place such that TeamCity can consume it and show in its logs) in case of error.

--- a/src/Paket.Core/Logging.fs
+++ b/src/Paket.Core/Logging.fs
@@ -1,29 +1,45 @@
 ﻿module Paket.Logging
 
 open System
+open System.IO
 
 let mutable verbose = false
+
+let mutable logFile : string option = None
+
+let setLogFile fileName =
+    let fi = FileInfo fileName
+    logFile <- Some fi.FullName
+    if fi.Exists then
+        fi.Delete()
+    else
+        if fi.Directory.Exists |> not then
+            fi.Directory.Create()
+
+let inline traceToFile (text:string) =
+    match logFile with
+    | Some fileName -> try File.AppendAllLines(fileName,[text]) with | _ -> ()
+    | _ -> ()
 
 /// [omit]
 let monitor = new Object()
 
 /// [omit]
-let inline tracen (s : string) = lock monitor (fun () -> printfn "%s" s)
+let inline tracen (s : string) = lock monitor (fun () -> traceToFile s; Console.WriteLine s)
 
 /// [omit]
 let inline tracefn fmt = Printf.ksprintf tracen fmt
 
 /// [omit]
-let inline trace (s : string) = lock monitor (fun () -> printf "%s" s)
+let inline trace (s : string) = lock monitor (fun () -> traceToFile s; Console.Write s)
 
 /// [omit]
 let inline tracef fmt = Printf.ksprintf trace fmt
 
-
 /// [omit]
 let inline traceVerbose (s : string) =
     if verbose then
-        lock monitor (fun () -> printfn "%s" s)
+        lock monitor (fun () -> traceToFile s; Console.WriteLine s)
 
 /// [omit]
 let inline verbosefn fmt = Printf.ksprintf traceVerbose fmt
@@ -34,7 +50,12 @@ let inline traceColored color (s: string) =
         (fun () ->
             let curColor = Console.ForegroundColor
             if curColor <> color then Console.ForegroundColor <- color
-            printfn "%s" s
+            traceToFile s
+            use textWriter = 
+                match color with
+                | ConsoleColor.Red -> Console.Error
+                | _ -> Console.Out
+            textWriter.WriteLine s
             if curColor <> color then Console.ForegroundColor <- curColor)
 
 /// [omit]

--- a/src/Paket.Core/PublicAPI.fs
+++ b/src/Paket.Core/PublicAPI.fs
@@ -49,7 +49,7 @@ type Dependencies(dependenciesFileName: string) =
     static member Create(path: string): Dependencies =
         let dependenciesFileName = Path.Combine(path,Constants.DependenciesFileName)
         if File.Exists dependenciesFileName then
-            Logging.tracef "%s already exists" dependenciesFileName
+            Logging.tracefn "%s already exists" dependenciesFileName
         else
             DependenciesFile(dependenciesFileName, InstallOptions.Default, [], [], []).Save()
         Dependencies(dependenciesFileName)

--- a/src/Paket/Paket.fsproj
+++ b/src/Paket/Paket.fsproj
@@ -28,7 +28,7 @@
     <DocumentationFile>
     </DocumentationFile>
     <StartArguments>update -f</StartArguments>
-    <StartArguments>install</StartArguments>
+    <StartArguments>install --log-file "C:\temp\log.txt"</StartArguments>
     <StartAction>Project</StartAction>
     <StartProgram>paket.exe</StartProgram>
     <StartWorkingDirectory>

--- a/src/Paket/Program.fs
+++ b/src/Paket/Program.fs
@@ -57,6 +57,7 @@ type CLIArguments =
     | [<AltCommandLine("--pre")>] Include_Prereleases
     | No_Auto_Restore
     | Creds_Migration of string
+    | Log_File of string
 with
     interface IArgParserTemplate with
         member s.Usage =
@@ -85,6 +86,7 @@ with
             | Nuget _ -> "allows to specify a nuget package."
             | Version _ -> "allows to specify a package version."
             | Creds_Migration _ -> "allows to specify credentials migration mode for convert-from-nuget."
+            | Log_File _ -> "allows to specify a log file."
             | FindRefs _ -> "finds all references to the given packages."
             | AddCredentials _ -> "add credentials to config file for the specified source."
 
@@ -110,6 +112,9 @@ let results =
             else Command.Unknown
         if results.Contains <@ CLIArguments.Verbose @> then
             verbose <- true
+
+        if results.Contains <@ CLIArguments.Log_File @> then
+            setLogFile <| results.GetResult <@ CLIArguments.Log_File @> 
 
         Some(command,results)
     with

